File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.py
import sys
input = sys.stdin.readline

def binary_search(start, end, target, res):
    while start < end:
        mid = (start + end) // 2
        if res[mid] < target:
            start = mid + 1
        else:
            end = mid
    return end

size = int(input()) 
A = list(map(int, input().split()))

result = [A[0]]
cnt = 1
for i in range(size):
    if result[cnt-1] < A[i]:
        result.append(A[i])
        cnt += 1
    else:
        idx = binary_search(0, cnt - 1, A[i], result)
        result[idx] = A[i]
        
print(cnt)

File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.cc
#include <iostream>
using namespace std;
#define MAXSIZE 1000001

int binary_search(int start, int end, int target, int res[]) {
    int mid;
    while(start < end) {
        mid = (start + end) / 2;
        if(res[mid] < target)
            start = mid + 1;
        else
            end = mid;
    }
    return end;
}

int main() {
    int size;
    int A[MAXSIZE] = {};
    cin >> size;
    for(int i = 0; i < size; i++)
        cin >> A[i];
    
    int result[MAXSIZE] = {};
    int cnt = 1;
    result[0] = A[0];
    for(int i = 0; i < size; i++) {
        if(result[cnt-1] < A[i]) {
            result[cnt] = A[i];
            cnt++;
        } else {
            int idx = binary_search(0, cnt-1, A[i], result);
            result[idx] = A[i];
        }
    }

    cout << cnt << endl;

    return 0;
}